pkg/reconciler.v1/common/interface.go

```diff
@@ -0,0 +1,248 @@
+
+	// ReconcilePods CAN be overridden if the logic to reconcile all Pods for the job changes.
+	ReconcilePods(
+		ctx context.Context,
+		job client.Object,
+		jobStatus *commonv1.JobStatus,
+		pods []*corev1.Pod,
+		rtype commonv1.ReplicaType,
+		spec *commonv1.ReplicaSpec,
+		replicas map[commonv1.ReplicaType]*commonv1.ReplicaSpec) error
+
+	// CreateNewPod CAN be overridden to customize how to create a new pod.
+	CreateNewPod(job client.Object, rt commonv1.ReplicaType, index string,
+		spec *commonv1.ReplicaSpec, masterRole bool, replicas map[commonv1.ReplicaType]*commonv1.ReplicaSpec) error
+
+	// DeletePod CAN be overridden to customize how to delete a pod of {name} in namespace {ns}.
+	DeletePod(ctx context.Context, ns string, name string) error
+
+	// DecoratePod CAN be overridden if customization to the pod is needed. The default implementation applies nothing
+	// to the pod.
+	DecoratePod(rtype commonv1.ReplicaType, podTemplate *corev1.PodTemplateSpec, job client.Object)
+}
+
+// ServiceInterface defines the abstract interface for Pod related actions, such like get, create or delete Service
+type ServiceInterface interface {
+	ReconcilerUtilInterface
+
+	// GetPortsFromJob CAN be overridden to customize how to find ports defined in the ReplicasSpec.
+	GetPortsFromJob(spec *commonv1.ReplicaSpec) (map[string]int32, error)
+
+	// GetServicesForJob CAN be overridden to customize how to find all services associated with this job.
+	GetServicesForJob(ctx context.Context, job client.Object) ([]*corev1.Service, error)
+
+	// FilterServicesForReplicaType CAN be overridden to customize how to filter out services for this Replica Type.
+	FilterServicesForReplicaType(services []*corev1.Service,
+		replicaType commonv1.ReplicaType) ([]*corev1.Service, error)
+
+	// GetServiceSlices CAN be overridden to customize how to generate service slices.
+	GetServiceSlices(services []*corev1.Service, replicas int, logger *logrus.Entry) [][]*corev1.Service
+
+	// ReconcileServices CAN be overridden to customize how to reconcile services for this job.
+	ReconcileServices(
+		job client.Object,
+		services []*corev1.Service,
+		rtype commonv1.ReplicaType,
+		spec *commonv1.ReplicaSpec) error
+
+	// CreateNewService CAN be overridden to customize how to create a new service.
+	CreateNewService(job metav1.Object, rtype commonv1.ReplicaType,
+		spec *commonv1.ReplicaSpec, index string) error
+
+	// DeleteService CAN be overridden to customize how to delete the service of {name} in namespace {ns}.
+	DeleteService(ns string, name string, job client.Object) error
+
+	// DecorateService CAN be overridden to customize this service right before being created
+	DecorateService(rtype commonv1.ReplicaType, podTemplate *corev1.PodTemplateSpec, job client.Object)
+}
+
+// JobInterface defines the abstract interface for Pod related actions, such like get, create or delete TFJob,
+// PyTorchJob or KFJob
+type JobInterface interface {
```

Comment:
Seems it brings more interfaces. I notice this change is very large and includes lots of refactoring. Maybe it's better to add more details in the issue for training leads to review? 

It would be great to cover following topics
1. What's the advantage of using `reconciler.v1` over `controller.v1`? 
2. What's the user facing changes?  for example, if user wants to build an operator, how this solution benefit their development? 

/cc @kubeflow/wg-training-leads

Reply:
For questions here, I updated https://github.com/kubeflow/common/pull/141#issue-687173348 to elaborate the benefit of `reconciler.v1`.

But from my perspective, I would keep both (reconciler and controller) for developers/users at different level.